File: cpp/leetcode/CrackSafe.cpp
//Leetcode Problem No 753 Cracking the Safe
//Solution written by Xuqiang Fang on 5 July, 2018

#include <iostream>
#include <vector>
#include <string>
#include <algorithm>
#include <unordered_map>
#include <unordered_set>
#include <stack>
#include <queue>

using namespace std;
class Solution{
public:
    string crackSafe(int n, int k){
        this->total_size = pow(k, n);
        string ans(n, '0');
        unordered_set<string> visited{ans};
        if(dfs(ans, n, k, visited)){
            return ans;
        }
        return "";
    }
private:
    int total_size;
    bool dfs(string& ans, const int n, const int k, unordered_set<string>& visited){
        if(visited.size() == total_size){
            return true;
        } 
        string node = ans.substr(ans.length()-n+1, n-1);
        for(char c='0'; c <'0'+k; ++c){
            node.push_back(c);
            if(!visited.count(node)){
                ans.push_back(c);
                visited.insert(node);
                if(dfs(, n, k, visited)){
                    return true;
                }
                visited.erase(node);
                ans.pop_back();
            }
            node.pop_back();
        }
        return false;
    }
};

int main(){
    Solution s;
    cout << s.crackSafe(2, 2) << endl;
    return 0;
}

File: py/leetcode/CrackSafe.py
'''
Leetcode problem No 753 Cracking the Safe
Solution written by Xuqiang Fang on 5 July, 2018
'''

class Solution(object):
    def crackSafe(self, n, k):
        """
        :type n: int
        :type k: int
        :rtype: str
        """
        N = pow(k, n)
        ans = [0 for i in range(n)]
        visited = set()
        visited.add(tuple(ans))
        self.dfs(ans, visited, n, k, N)
        return ''.join([str(i) for i in ans]) 

    def dfs(self, ans, v, n, k, N):
        if len(v) == N:
            return True
        for i in range(k):
            ans.append(i)
            t = tuple(ans[-n:])
            if t not in v:
                v.add(t)
                if self.dfs(ans, v, n, k, N):
                    return True
                v.remove(t)
            ans.pop()
        return False

def main():
    s = Solution()
    print(s.crackSafe(3,8))
main()
